### PR TITLE
Refactor logging

### LINKED ISSUE #117: Revisit verbosity (stdout/stderr)
In https://github.com/in-toto/in-toto/issues/6 we discussed and fixed many initial user feedback issues.
In the course of a layout tool user testing session, it was noted that when looking at stdout/stderr of the used tool (`in-toto-mock`) it was unclear if the command was executed properly or not.

Here are some general user feedback related items open for discussion and fixing:
- Make better use of verbosity levels
- Reduce the output if `-v` flag is not specified
- Make clear if a command failed or not
- Colorize output
- Do something about `subprocess32` warning (seems to confuse users)

### PR DESCRIPTION
**Pull request by lukpueh**
**Fixes issue #**:
Closes #117 

**Description of the changes being introduced by the pull request**:
The goal of this PR is to make better use of Python's `logging` module for debugging as well as user feedback.

Before, in-toto used wrappers for logging functions directly on the `in_toto.log` module. Now these wrappers are removed and instead we register a base `in_toto` logger, using a custom `logging.Logger` subclass with an enhanced error method and a convenience method to set the logging level, e.g. in command line interfaces that accept a `verbose` and/or `quiet` option, which are also added in this PR.

See  in_toto.log's doc header for detailed usage instructions.

The PR also removes the `LOG_LEVEL` setting and adds a `DEBUG` setting (boolean) which affects the format and log level of the in-toto base logger.

The PR also replaces some `warning` with `info` log statements. For user feedback we should always use log type info, because it is disabled by default when using the library directly, but enabled when using command line interfaces with the `verbose` option.

Furthermore the PR fixes test log inundation by
- replacing the base logger's  default `StreamHandler` with `NullHandler` when running tests to suppress all `logging` based output.
- adding global flag buffer flag to runtests' `TextTestRunner` to suppress non `logging` based output.
- adding a stderr Pipe to subprocess calls in `in_toto.gpg`, to suppress gpg warnings (we are not
interested in stderr (but stdout) to parse keys and stderr only floods the test log due to permission warnings related to the test gpg keychain.

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

### CODE REVIEW

File: in_toto/in_toto_mock.py [hunk 1]
```diff
@@ -91,8 +96,8 @@ def main():
 
   args = parser.parse_args()
 
-  # Turn on all the `log.info()` in the library
-  log.logging.getLogger().setLevel(log.logging.INFO)
+  # Default to verbose
```
Comment (SantiagoTorres) [inline in_toto/in_toto_mock.py]: I don't know how descriptive this comment is. I personally find it a little confusing.

File: in_toto/log.py [hunk 2]
```diff
@@ -12,30 +12,118 @@
...
+  def setLevelVerboseOrQuiet(self, verbose, quiet):
+    """Convenience method to set the logger's verbosity level based on the
+    passed booleans verbose and quiet (useful for cli tools). """
+    if verbose:
+      self.setLevel(logging.INFO)
+
+    elif quiet:
+      # TODO: Is it enough to use logging.CRITICAL + 1 to suppress all output?
+      # A saver way would be to use a NullHandler or Filters.
+      self.setLevel(self.QUIET)
```
Comment (SantiagoTorres) [inline in_toto/log.py:105]: I think this is "ok" imho.

File: tests/test_in_toto_record.py [hunk 3]
```diff
@@ -105,11 +101,6 @@ def test_start_stop(self):
     self._test_cli_sys_exit(["stop"] + args + ["--products",
         self.test_artifact2, self.test_artifact2], 0)
 
-    # Start/stop recording verbosely
-    args = ["--step-name", "test4", "--key", self.key_path, "--verbose"]
-    self._test_cli_sys_exit(["start"] + args, 0)
-    self._test_cli_sys_exit(["stop"] + args, 0)
-
```
Comment (SantiagoTorres) [inline tests/test_in_toto_record.py:112]: Why did this test disappear?

File: tests/test_in_toto_run.py [hunk 4]
```diff
@@ -38,8 +37,7 @@
 
 import in_toto.gpg.util
 
-# Suppress all the user feedback that we print using a base logger
-logging.getLogger().setLevel(logging.CRITICAL)
+
```
Comment (SantiagoTorres) [inline tests/test_in_toto_run.py:40]: I think this whitespace is one too many.

File: tests/test_settings.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+import in_toto.settings
+
+
+class TestSettings(unittest.TestCase):
+  def test_debug_not_true(self):
+    """in_toto.settings.DEBUG should not be commited with True. """
+    self.assertFalse(in_toto.settings.DEBUG)
+
+if __name__ == "__main__":
+  unittest.main()
```
Comment (SantiagoTorres) [inline tests/test_settings.py:28]: What's the difference between `test_settings` and `test_user_settings`?

**Review by SantiagoTorres — APPROVED** [review]
LGTM, although I have some questions.

File: in_toto/in_toto_mock.py (hunk 1 above)
Comment (lukpueh) [reply in_toto/in_toto_mock.py]: I can rephrase, e.g.:
*in-toto-mock should not be used to secure the supply chain but only to try out in-toto-run, hence it gives verbose user feedback*

File: in_toto/log.py (hunk 2 above)
Comment (lukpueh) [reply in_toto/log.py:105]: :)

File: tests/test_in_toto_record.py (hunk 3 above)
Comment (lukpueh) [reply tests/test_in_toto_record.py:112]: The test used to trigger an `if` statement in in_toto_record. That `if` statement was replaced by a logger method call `setLevelVerboseOrQuiet`, which I test separately in `test_log.py`.

Therefore, we are good in terms of coverage. In terms of interface testing, you are right, I could call the the cli tool with the verbose flag and test if it passes the right thing to `setLevelVerboseOrQuiet`. Not sure if necessary. What do you think?

File: tests/test_in_toto_run.py (hunk 4 above)
Comment (lukpueh) [reply tests/test_in_toto_run.py:40]: This is where our code style guideline actually suggests [5 blank lines](https://github.com/secure-systems-lab/code-style-guidelines#blank-lines), which I find too much. I think 3 are a fair compromise...

File: tests/test_settings.py (hunk 5 above)
Comment (lukpueh) [reply tests/test_settings.py:28]: One tests the `settings` module and the other one tests the `user_settings` module. The latter can be used to configure the former by loading environment variables and rcfiles.

File: tests/test_in_toto_run.py (hunk 4 above)
Comment (SantiagoTorres) [reply tests/test_in_toto_run.py:40]: wow, TIL. Ok I guess...

File: tests/test_settings.py (hunk 5 above)
Comment (SantiagoTorres) [reply tests/test_settings.py:28]: > One tests the settings module and the other one tests the user_settings module

I mean that much I could tell. I was imagining both of them are the same functionality, with `user_settings` being just a module/function to override the former and thus needed no conceptual separation. 

I don't think it's a deal at all though.

File: tests/test_in_toto_record.py (hunk 3 above)
Comment (SantiagoTorres) [reply tests/test_in_toto_record.py:112]: I don't think it's necessary. I just wanted to make sure you didn't delete that test for the sake of testing or something akin to that.

**Comment by lukpueh** [comment]
Thanks for the reviewing and the button clicking, @SantiagoTorres

**Pull request merged**